pkg/controller/certificatesigningrequests/acme/acme.go

```diff
@@ -0,0 +1,306 @@
+		// are cluster scoped and so hashes won't conflict. This is likely from
+		// someone manually creating the order out of band. We can only error.
+		return errors.New("found Order resource not owned by this CertificateSigningRequest, retrying")
+	}
+
+	log = logf.WithRelatedResource(log, order)
+
+	// If the acme order has failed then so too does the
+	// CertificateSigningRequest meet the same fate.
+	if acme.IsFailureState(order.Status.State) {
+		err := fmt.Errorf("order is in %q state: %s", order.Status.State, order.Status.Reason)
+		message := fmt.Sprintf("Failed to wait for order resource %s/%s to become ready: %s", expectedOrder.Namespace, expectedOrder.Name, err)
+
+		a.recorder.Event(csr, corev1.EventTypeWarning, "OrderFailed", message)
+		ctrlutil.CertificateSigningRequestSetFailed(csr, "OrderFailed", message)
+		_, uerr := a.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return uerr
+	}
+
+	if order.Status.State != cmacme.Valid {
+		a.recorder.Event(csr, corev1.EventTypeNormal, "OrderPending",
+			fmt.Sprintf("Waiting on certificate issuance from order %s/%s: %q",
+				expectedOrder.Namespace, order.Name, order.Status.State))
+
+		log.V(logf.DebugLevel).Info("acme Order resource is not in a ready state, waiting...")
+		return nil
+	}
+
+	if len(order.Status.Certificate) == 0 {
+		a.recorder.Event(csr, corev1.EventTypeNormal, "OrderPending",
+			fmt.Sprintf("Waiting for order-controller to add certificate data to Order %s/%s",
+				expectedOrder.Namespace, order.Name))
+
+		log.V(logf.DebugLevel).Info("order controller has not added certificate data to the Order, waiting...")
+		return nil
+	}
+
+	x509Cert, err := pki.DecodeX509CertificateBytes(order.Status.Certificate)
+	if err != nil {
+		message := fmt.Sprintf("Deleting Order with bad certificate: %s", err)
+		a.recorder.Event(csr, corev1.EventTypeWarning, "OrderBadCertificate", message)
+		log.Error(err, "failed to decode x509 certificate data on Order resource.")
+		// Deleting the order here will cause a re-sync since the Order is owned by
+		// this CertificateSigningRequest.
+		return a.acmeClientV.Orders(order.Namespace).Delete(ctx, order.Name, metav1.DeleteOptions{})
+	}
+
+	if ok, err := pki.PublicKeyMatchesCertificate(req.PublicKey, x509Cert); err != nil || !ok {
+		a.recorder.Event(csr, corev1.EventTypeWarning, "OrderBadCertificate", "Deleting Order as the signed certificate's key does not match the request")
+		log.Error(err, "The public key in Order.Status.Certificate does not match the public key in CertificateSigningRequest.Spec.Request. Deleting the order.")
+		// Deleting the order here will cause a re-sync since the Order is owned by
+		// this CertificateSigningRequest.
+		return a.acmeClientV.Orders(order.Namespace).Delete(ctx, order.Name, metav1.DeleteOptions{})
+	}
+
+	// Update the status.certificate first so that the sync from updating will
+	// not cause another issuance before setting the CA.
```

Comment:
question: what do you mean by "first"? There is no other Update/UpdateStatus after this, I do not understand the comment 🤔

Reply:
Old comment that should be removed.